### ISSUE TITLE
Issues while trying to reverse engineer an AppImage, then "fix" and rebuild it

### ISSUE THREAD

**Issue by treoleopi** [AppImage/AppImageKit#761]
My issue is (most likely) due to problems with the AppRun file. 

Some guy has done a great job here: 

https://github.com/davidtavarez/portable-monero-miners

But unfortunately this project is no longer useful since Monero changed their code radically on April 6th and while the executable loads, it doesn't do the job it was defigned for...

So anyway, I downloaded and extracted the Appimage, then tried replacing the executable inside with the newer exe but then get errors that a shared lib can't be found... I tried making my own AppImage the n but then get execv errors that "no such file or directory"...doing cat on the AppImage does't help since it seems to be binary?

Any thoughts? how can I open the AppRun file...also: do I need the source code at all or can't I just have the libs + exe in the directory? thanks!

**Comment by treoleopi** [comment]
Ps:
Sorry I mean't "cat" on the AppRun file, not AppImage...

**Comment by TheAssassin** [comment]
Well, obviously, replacing some executable with another without fixing its dependencies as well is a horrible idea...

You need to provide _way_ more information. What exactly do you want to know? You're not asking a concrete question. "I get some errors" without posting an _exact_ message or, even better, a full log is not helping anybody in helping you. Do you suppose us to "guess" your problem? That's impossible without data.
Next, AppImages are built by third parties, they control what goes into them, all of them are built in a really different way, and we could only guess what's going wrong with just those few lines you wrote with almost no information other than "some error message".

"Do I need the source code?" No, you don't, but if the app isn't open source, you are often not allowed to re-package it, etc. Such a general reply won't help you. There's ways to bundle existing libraries, but I'm not sure whether that's what you want to do.

Let me ask you directly: did you try to talk to this project? We generally can't provide much support for third party AppImages. We can only try to help for concrete problems. Please describe your exact problem, maybe upload an AppImage that doesn't work. And if you didn't try to talk to this project you linked to, please try that first. They built your AppImage, and they can help you fix it best. If they don't react, you can come back here, and we'll try to help you.

**Comment by treoleopi** [comment]
OK, I am now trying everything from scratch...will post exact info later, the 'portable' AppImage now btw seems too big ~40MB vs 3MB the original one but that's because I included the whole /lib dir from my distribution just to make sure libs will not be missed. You are correct I know very little about it since I did't know about AppImages up untill 3/4 hours ago but they seem very cool idea indeed!

**Comment by TheAssassin** [comment]
@treoleopi I recommend you to join us on IRC, that'll make helping you a lot easier.

**Comment by probonopd** [comment]
Also, is this a question related to AppImageKit, or a question related to the Monero AppImage? We can help with the former but possibly less so with the latter...

**Comment by treoleopi** [comment]
Sorrry something more specific, I am getting "execv: cannot find file or directory", using an AppRun like this:


==================================
#!/bin/bash
HERE="$(dirname "$(readlink -f "${0}")")"
export LD_LIBRARY_PATH=${HERE}/usr/lib/x86_64-linux-gnu/:$HERE/usr/lib/:${HERE}/lib/
export PATH=${HERE}/usr/sbin:${HERE}/usr/bin:/bin:/sbin:$PATH
"${HERE}"/usr/bin/xmrig "$@"
=================================

So it's unable to find the file: "xmrig.Appdir/usr/bin/xmrig"? 
xmrig is in the Appdir, chmod 777, executable tested....10x.

**Comment by probonopd** [comment]
`AppRun` itself is in the root of the AppDir, chmod 777, executable tested?

**Comment by probonopd** [comment]
Does the application run when you execute `AppRun`? (Not inside an AppImage)

**Comment by TheAssassin** [comment]
I updated your comment to add code tags (check the `<>` button above the editor, or use ` ``` code ``` ` directly).

The following line's quoting is totally broken by the way:
```
HERE="$(dirname "$(readlink -f "${0}")")"
```

Use:

```
HERE=$(dirname $(readlink -f "$0"))
```

`$()` doesn't need quotes, only the `$0` argument needs some, as it might contain spaces, which could be interpreted by bash. With quotes, they are not.

```
my    path.AppImage
```
might become

```
my path.AppImage
```

otherwise. `$(...)` however is treated as a single unit, and won't be interpreted by bash, therefore doesn't need any kind of quotes.

Also, you should use:

```
exec "$HERE"/usr/bin/xmrig "$@"
```

The `exec` makes bash exit and replace itself with the command.

Also, you shouldn't add non-AppImage directories to the `$PATH`. And `$HERE` could use some escaping, for the above reasons.

TL;DR:

```sh
#!/bin/bash
HERE=$(dirname $(readlink -f "$0"))
export LD_LIBRARY_PATH="$HERE"/usr/lib/x86_64-linux-gnu/:"$HERE"/usr/lib/:"$HERE"/lib/
export PATH="$HERE"/usr/sbin:"$HERE"/usr/bin:"$PATH"
"$HERE"/usr/bin/xmrig "$@"
```

P.S.: You could just call `xmrig` without the absolute path as well. It's in your PATH after the second last line.

**Comment by treoleopi** [comment]
@probonopd Yes, the app seems to be running when I try say ./AppRun while just cd-ing to the directory.

I will now try to call the exe directly, upon what the @TheAssassin suggested, will keep you guys posted and try uploading .zip if successfull...

**Comment by TheAssassin** [comment]
You can upload it to https://transfer.sh without zipping it.

**Comment by probonopd** [comment]
You need to be able to run it without cd'ing to the directory. E.g.,

```
cd Your.AppDir
cd ../
./Your.AppDir/AppRun
```

If `AppRun` only works if you are in the same directory, then something is wrong with the `AppRun` or the application.

**Comment by treoleopi** [comment]
I am now trying to compile on old Ubuntu 12...this is "fun"...

**Comment by probonopd** [comment]
> Is it possible to update AppRun to include Fuse/libfuse?

I don't think so and even if it was possible, you'd still need a working FUSE setup (with `/dev/fuse` and correct permissions) which a system is unlikely to have if libfuse is not there. But hey, on systems without FUSE the user can still run `--appimage-extract`.

> some configurations complain that (...) Glibc can't be found?

I think you are getting errors when you are trying to run an AppImage on a system older than the one you used for building.

**Comment by probonopd** [comment]
> I am now trying to compile on old Ubuntu 12...this is "fun"...

The oldest still-supported (by Canonical) Ubuntu version is 14.04.

**Comment by treoleopi** [comment]
Yeah...I've compiled on 14 before...I've been struggling on 12 for like 2 hours now, I had to do like 20+ changes but still things are failing (update reps, add libuv, add libmicrohttp, update g++, cmake...). 

I may do it again on 14 anyway again, it should be way simpler. 

Also, on one Debian install on my VM I got "core dumped"? Is this again related to older version vs newer AppImageversion problem?

**Comment by Calinou** [comment]
> If someone uses --appimage-extract on a system without Fuse, does this mean the executable will still run "like an AppImage" in the sense that it will again check only its directory for libraries? I mean, it's not mandatory for a system to have required libs installed, correct?

Yes, as long as the `AppRun` file is run (and not the binary directly).

**Comment by treoleopi** [comment]
@Calinou:

Yes, sorry, I meant if someone clicks on the AppRun...it makes sense though :).

**Comment by TheAssassin** [comment]
> Yes, as long as the AppRun file is run (instead of directly running the binary). The AppRun file sets environment variables such as LD_LIBRARY_PATH so that the binary looks for libraries in the right directories.

... and as long as the app doesn't rely on AppImage runtime features, i.e., `$APPIMAGE` or `$APPDIR`. But 99% will run.

**Comment by azubieta** [comment]
@treoleopi any updates on this topic?

**Comment by TheAssassin** [comment]
One of the last comments contained "last question", so I consider this to be finished. Closing as resolved.
